charts/fluent-bit/templates/ingress.yaml

```diff
@@ -0,0 +1,41 @@
+{{- if .Values.ingress.enabled -}}
```

Comment:
Cannot be exact same, need some changes, because ingress should be able to expose extra ports eg.  http input, but I changed it according latest helm.

Reply:
Thanks I see this was done via extraHosts, there is another viable approach for this: https://github.com/jaegertracing/helm-charts/pull/308/files

not sure if one is better than the other though, @stevehipwell WDYT?

Reply:
What exactly are you trying to do here? Ingress is a layer on top of service so the logic should be in service values which ingress can then implement. Also please don't use `ImplementationSpecific` as the default `pathType`, it breaks some ingress controllers, `Prefix` is the best default and in a lot of cases can be hardcoded.

This is my ingress pattern for multiple service ports; you'll find it or the single port version in my charts plus the kube-prometheus-stack & Grafana charts.

https://github.com/stevehipwell/helm-charts/blob/master/charts/nexus3/templates/ingress.yaml

Reply:
@stevehipwell please look at service implementation, there is `service.port`, which according default values expose "statistics" `HTTP_Server` https://github.com/fluent/helm-charts/blob/main/charts/fluent-bit/values.yaml#L206 and there are `extraPorts`. 
What I want to provide is a way how to access extraPorts via ingress. In this implementation, you can put ingress on top of `service.port` or `extraPorts` or just `extraPorts` or both. I took same naming convention and used extraHosts.
Regarding `ImplementationSpecific`, it is taken from https://github.com/helm/helm/blob/d4cc130fa90027e8c1c69ad2a06c6dee4c31d29d/pkg/chartutil/create.go#L158
as was suggested by @naseemkullah, that part of code is used by "helm create ( create a new chart with the given name)", so maybe the helm guys are wrong...and it's not default, it's commented out as example how to fill values, so it won't break anything, but I have no problem to put there `Prefix` as example.
Regarding jaegertracing approach, this one is more straightforward, mixing types in one value just complicates things.

Reply:
@tombokombo my point is that if ports have been added to the service then the ingress should be exposing them, you shouldn't need additional complexity/config for this. There is a case for known ports such as metrics to be exposed by the service but not on the ingress but this requires prior knowledge of what's running.

I'm also not sure what the purpose of a configurable `path` or `pathType` is in the context of this chart?

For me the ingress values should be as below as all of the rest of the configuration is static or in another value such as `service` or `extraPorts`.

```yaml
ingress:
  enabled: false
  className: ""
  annotations: {}
    # kubernetes.io/ingress.class: nginx
    # kubernetes.io/tls-acme: "true"
  hosts: []
  # - fluent-bit.example.tld
  tls: []
  #  - secretName: fluent-bit-example-tld
  #    hosts:
  #      - fluent-bit.example.tld
```

Reply:
I think the issue is that since fluent bit is very flexible there may be ports (other than the HTTP Service) that one wants to expose via ingress and others just via clusterIP - extraHosts is a way to achieve this granularity, offloading  that decision making to the user, and in a lot of cases for fluent-bit this is whats required since its so flexible and I've opted to maintain the flexibility (which does add extra work for the chart consumer) instead of reflect in values the many possible combinations into the chart (see any values that start with `extra`).

My only concern was whether the jaeger approach was better or not. I'll approve this since there seems to be no obviously better way to achieve the flexibility.

That said I also see the point of not having to add extra logic, but this limits what we can do - though the limit may be ok which would be that all extra ports are exposed via a boolean `exposeExtraPorts` or similarly by creating a new value list called `exposedExtraPorts` - @tombokombo is your use case to expose all the extra ports? do you need them on different hosts?

re: aggregator chart I'm more keen towards something like an `aggregator-example-values.yaml` in the README - maintaining another chart is also its own kind of cognitive load and I think we'd face the same issues we are facing here minus the code reuse (unless we moved common code to a library chart but that seems expensive). But you can check with @edsiper ultimately.

Reply:
So I've checked all inputs, for now it doesn't make sense to define paths as was suggested by @stevehipwell, so I've removed this layer of complexity, but we still need to define port in extraHosts.
```yaml
  hosts: []
  # - host: fluent-bit.example.tld
  extraHosts: []
  # - host: fluent-bit-extra.example.tld
      ## specify extraPort number
  #   port: 5170
  tls: []
  #  - secretName: fluent-bit-example-tld
  #    hosts:
  #      - fluent-bit.example.tld
```
@naseemkullah I don't need "statistics" HTTP_Server exposed via ingress at all, I just need http input to be under one host, so my use case is one extraPort.

Actually, I can imagine more complex use-cases for example public domain/host http input exposed with basic auth (achieved via ingress annotations) under one controller and internal domain without auth under other controller with different class name. This case would need range over ingresses...I've just want to say that there are really lot of cases, which will not be handled.

But back to this problem, I'm thinking that if you choose to disable HTTP_Server by default, there will be no need for `service.port` and everything could be handled by `service.ports` with no extra. If user decides to enable it, he can map it's port to service and ingress as before and similarly with other inputs, but this approach will need few changes.

Reply:
> But back to this problem, I'm thinking that if you choose to disable HTTP_Server by default, there will be no need for service.port and everything could be handled by service.ports with no extra. If user decides to enable it, he can map it's port to service and ingress as before and similarly with other inputs, but this approach will need few changes.

Can you provide some YAML as to what you mean please?

Reply:
Sure, something like this
```yaml
config:
  service: |
    [SERVICE]
        Daemon Off
        Flush 1
        Log_Level {{ .Values.logLevel }}
        Parsers_File parsers.conf
        Parsers_File custom_parsers.conf
        HTTP_Server Off
        # HTTP_Listen 0.0.0.0
        # HTTP_Port 2020
        Health_Check On
  inputs: |
     [INPUT]
        Name http
        host 0.0.0.0
        port 80
...
...
service:
  type: ClusterIP
  labels: {}
  # nodePorts: 
  # -30020
  annotations: {}
#   prometheus.io/path: "/api/v1/metrics/prometheus"
#   prometheus.io/port: "2020"
#   prometheus.io/scrape: "true"
    ports:
    # - port: 2020
    #   containerPort: 2020
    #   protocol: TCP
    #   name: http
    - port: 80
      containerPort: 80
      protocol: TCP
      name: httpin

ingress:
  enabled: true
  hosts:
  # - host: fluent-bit.example.tld
  #   port: 2020
  - host: fluent-bit-httpin.example.tld
    port: 80
  tls: []
  #  - secretName: fluent-bit-example-tld
  #    hosts:
  #      - fluent-bit.example.tld

Reply:
What are the actual use cases the ingress changes are trying to cover? Would it not be best to separate the HPA & PDB changes as these are a lot simpler and idiomatic?

Reply:
Will send another PR just with HPA & PDB.
According fluent-bit functionality, ingress should be able to expose http input and HTTP_Server (AFAIK).

Reply:
PR for hpa and pdb https://github.com/fluent/helm-charts/pull/182

Reply:
Hi, regarding ingress, could we agree on latest version without paths?

Reply:
@tombokombo don't you need to remove the HPA & PDB code from this PR and rename it to ingress only?

Reply:
@stevehipwell could we move with this feature?

Reply:
@tombokombo I'd like to see the API selection and other conditions moved into helper functions to lower the cognitive complexity in the `Ingress` manifest. I've shared an example above from one of my charts, but the Grafana chart ingress and the ingresses in the kube-prometheus-stack charts also use the same fundamental pattern.

Reply:
@stevehipwell done, logic moved to helper